core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java

```diff
@@ -66,6 +103,138 @@ public void checkRead(Object key) {
+      if (trace) {
+         log.tracef("Added partially rollback transaction %s", globalTransaction);
+      }
+      partialTransactions.put(globalTransaction, new RollbackTransactionInfo(globalTransaction, affectedNodes, lockedKeys));
+      return true;
+   }
+
+   @Override
+   public boolean addPartialCommit2PCTransaction(GlobalTransaction globalTransaction, Collection<Address> affectedNodes,
+                                                 Collection<Object> lockedKeys, EntryVersionsMap newVersions) {
+      if (trace) {
+         log.tracef("Added partially committed (2PC) transaction %s", globalTransaction);
+      }
+      partialTransactions.put(globalTransaction, new Commit2PCTransactionInfo(globalTransaction, affectedNodes,
+                                                                              lockedKeys, newVersions));
+      return true;
+   }
+
+   @Override
+   public boolean addPartialCommit1PCTransaction(GlobalTransaction globalTransaction, Collection<Address> affectedNodes,
+                                                 Collection<Object> lockedKeys, List<WriteCommand> modifications) {
+      if (trace) {
+         log.tracef("Added partially committed (1PC) transaction %s", globalTransaction);
+      }
+      partialTransactions.put(globalTransaction, new Commit1PCTransactionInfo(globalTransaction, affectedNodes,
+                                                                              lockedKeys, modifications));
+      return true;
+   }
+
+   @Override
+   public boolean isTransactionPartiallyRolledBack(GlobalTransaction globalTransaction) {
+      TransactionInfo transactionInfo = partialTransactions.get(globalTransaction);
+      if (trace) {
+         log.tracef("Can release resources for transaction %s. Transaction info=%s", globalTransaction, transactionInfo);
+      }
+      return transactionInfo == null || transactionInfo.isRolledBack(); //if we are going to commit, we can't release the resources yet
+   }
+
+   @Override
+   public Collection<GlobalTransaction> getPartialTransactions() {
+      return Collections.unmodifiableCollection(partialTransactions.keySet());
+   }
+
+   @Override
+   public boolean canRollbackTransactionAfterOriginatorLeave(GlobalTransaction globalTransaction) {
+      boolean canRollback = availabilityMode == AvailabilityMode.AVAILABLE &&
+            !getLastStableTopology().getActualMembers().contains(globalTransaction.getAddress());
+      if (trace) {
+         log.tracef("Can rollback transaction? %s", canRollback);
+      }
+      return canRollback;
+   }
+
+   @Override
+   public void onTopologyUpdate(CacheTopology cacheTopology) {
+      boolean isStable = isTopologyStable(cacheTopology);
+      if (trace) {
+         log.tracef("On stable topology update. Has pending tx? %b. Is stable? %b. topology=%s",
+                    !partialTransactions.isEmpty(), isStable, cacheTopology);
+      }
+      if (isStable) {
```

Comment:
I think you need to do this when you get a new stable topology as well. Probably not for "regular" merges, but if the admin decides to force the cluster to become available, you'll get the regular topology update first and the stable topology update later.